Prepare leaking GC for multithreading

aykevl opened the pull request:
This makes the leaking GC (`-gc=leaking`) ready for multithreading. The leaking GC is the easiest real heap allocator to add support for, since it can increment the heap pointer with a simple atomic add operation.
This PR does not actually make the leaking GC support multithreading, but it makes it very easy to do so in the future simply by changing the new llsync package (see https://github.com/tinygo-org/tinygo/pull/4559).

This is the first PR of (hopefully) a bunch to follow that will gradually add support for multithreading to various parts of TinyGo. I split it in many pieces to make it hopefully easier to review.

This PR is intended to be merged by rebasing.

aykevl committed:
internal/task: implement PMutex

PMutex is a mutex when threading is possible, and a dummy mutex-like
object (that doesn't do anything) otherwise.

aykevl commented:
Note: not super convinced about the "llsync" name (meaning "low level sync"). Others I have considered: "sync" (can be confused by the stdlib sync package), "atomic" (same), "futex" (there's more in there than just futexes), "mutex" (it's more than mutexes), "internalsync" (too long), etc.
Please let me know if you can think of a better name!

dgryski commented:
`tinysync` ?

aykevl commented:
Actually I realized I'd like to be able to use these synchronization primitives inside the internal/task package and some synchronization primitives (such as the pseudo-futex for the non-threaded scheduler) need internal/task. So that's going to be a circular dependency. Therefore I've moved the new stuff into the internal/task package as well, to simplify things.
Also, putting them there does make some sense: these primitives depend on which kind of scheduler (and therefore which kind of task) is in use.

This nicely sidesteps the naming question :)

aykevl commented:
Can someone take a look at this PR?

eliasnaur commented:
Not a maintainer, but LGTM. The changes are fairly harmless: adding the passthrough `pseudoAtomic` type and refactoring the leaking GC to use it.

dgryski commented:
I think there's more than just sprinkling a few atomics onto some integers to truly make this thread-safe.  For example, the growing logic and checking `heapEnd` needs to be similarly protected.   Generally to make things thread-safe without wrapping the whole thing in a mutex means a lot more atomic test-and-set loops.  I can try to take a look at this next week.

aykevl commented:
> For example, the growing logic and checking `heapEnd` needs to be similarly protected.

Oops, you are correct! I believe it's correct for a non-growing heap, but that's only the case on microcontrollers.

aykevl committed:
runtime: prepare the leaking GC for concurrent operations

This uses the task.PMutex parallel-only-mutex type to make the leaking
GC parallelism safe. The task.PMutex type is currently a no-op but will
become a real mutex once we add true parallelism.

aykevl commented:
Updated the PR. With `growHeap`, there doesn't seem to be a trivial way to use atomic operations so I've used a lock instead.

dgryski commented on src/runtime/gc_leaking.go:
```diff
@@ -43,6 +48,8 @@ func alloc(size uintptr, layout unsafe.Pointer) unsafe.Pointer {
 		// Failed to make the heap bigger, so we must really be out of memory.
 		runtimePanic("out of memory")
```
Unlock mutex before throwing `panic`?

aykevl replied:
It's a runtime panic, which cannot be recovered from. So unlocking the mutex won't help anybody.

deadprogram commented:
Any other feedback on this PR @dgryski ?

dgryski reviewed (approved):
LGTM.
